astropy/utils/decorators.py

```diff
@@ -863,7 +866,212 @@ def __get__(self, obj, objtype=None):
+    ... except AttributeError as e:
+    ...     print(e)
+    'Example' object has no attribute 'identify'
+
+    Likewise, ``identify`` is not in the :func:`dir` of instances of
+    ``Example``.
+
+    >>> print("identify" in dir(Example()))
+    False
+
+    `on_metaclass` can also be used to promote a `classmethod` to a
+    `classmethod` on the metaclass, making a metaclass-classmethod.
+
+    >>> class Example(metaclass=ExampleMeta):
+    ...     @on_metaclass
+    ...     @classmethod
+    ...     def identify(mcls):
+    ...         print('this overrides the metaclass method for {}'.format(mcls))
+    ...
+    >>> Example.identify()
+    this overrides the metaclass method for <class 'astropy.utils.decorators.ExampleMeta'>
+
+    `on_metaclass` also works with `property` to make a class-level property.
+
+    >>> class Example(metaclass=ExampleMeta):
+    ...     @on_metaclass
+    ...     @property
+    ...     def identify(cls):
+    ...         print('this overrides the metaclass method for {}'.format(cls))
+    ...
+    >>> Example.identify
+    this overrides the metaclass method for <class 'astropy.utils.decorators.Example'>
+
+    To make a metaclass-level property, stack a `on_metaclass` with a
+    `classmethod` and a `property`.
+
+    .. code-block::
+
+        class Example(metaclass=ExampleMeta):
+            @on_metaclass
+            @classmethod
+            @property
+            def identify(mcls):
+                print('this overrides the metaclass method for {}'.format(mcls))
+
+    Notes
+    -----
+    Note this must be the top-most decorator in a decorator stack and MUST be
+    used as a decorator s.t. it calls ``__set_name__``.
+
+    Methods decorated with ``on_metaclass`` are not actually part of the
+    metaclass but only emulate their behavior. Consequently it is necessary to
+    modify the enclosing class, replacing the ``__dir__`` method. Meta-class
+    methods are in the :func:`dir` of classes, but not their instances, so the
+    ``__dir__` method is replaced by a descriptor that can filter out select
+    entries. For code details see the ``__set_name__`` method of this class.
+    """
+
+    _attr_name: str
+
+    def __set_name__(self, encl_cls, attr_name: str) -> None:
+        # The wrapped attrs / method
+        self._attr_name = attr_name
+
+        # Make the ``__dir__`` filterable, if not already.
+        if not isinstance(vars(encl_cls).get("__dir__"), _dir_):
+            encl_cls.__dir__ = _dir_(encl_cls.__dir__)
+
+        # Add this method to the ``__dir__`` filter.
+        vars(encl_cls)["__dir__"].exclude.add(attr_name)
+
+    def __get__(self, encl_inst: Optional[object], encl_cls: Optional[type]=None):
+        # @classmethod, but restricting to getting from `encl_cls`
+        if encl_inst is not None:
+            msg = (f"{encl_inst.__class__.__name__!r} object "
+                   f"has no attribute {self._attr_name!r}")
+            raise AttributeError(msg)
```

Comment:
It might be useful to have distinct error messages depending on whether the attribute can be accessed from the class or not.